toqito/matrices/tests/test_operator_sinkhorn.py

```diff
@@ -0,0 +1,182 @@
+# 4*3*2 != 8. So dim array should not match
+dim_invalid_dimarr = [4, 3, 2]
+expected_msg_invalid_dimarr = " ".join(
+    (f"Product of dimensions {dim_invalid_dimarr}", f"does not match rho dimension {rho_invalid_dimarr.shape[0]}.")
+)
+
+"""Test operator sinkhorn on non-square input matrix."""
+rho_nonsquare = np.random.rand(4, 5)
+dim_nonsquare = [2, 2]
+expected_msg_nonsquare = "Input 'rho' must be a square matrix."
+
+
+params1 = [
+    ([rho_singular, dim_singular], expected_msg_singular),  # singular matrix
+    ([rho_invalid_sdim, dim_invalid_sdim], expected_msg_invalid_sdim),  # invalid single dim
+    ([rho_invalid_dimarr, dim_invalid_dimarr], expected_msg_invalid_dimarr),  # invalid dim array
+    ([rho_nonsquare, dim_nonsquare], expected_msg_nonsquare),  # non square matrix input
+]
+
+
+@pytest.mark.parametrize("test_input,expected_msg", params1)
+def test_operator_sinkhorn_errors(test_input, expected_msg):
+    """Test Error handling in different scenarios."""
+    # unpack test inputs
+    rho, dim1 = test_input
+
+    try:
+        operator_sinkhorn(rho, dim=dim1)
+    except ValueError as e:
+        assert str(e) == expected_msg
+
+
+def test_operator_sinkhorn_max_iterations():
+    """Test operator sinkhorn on insufficient iteration limit."""
+    # function should raise a RuntimeError
+
+    rho_random = random_density_matrix(4, seed=42)
+    try:
+        operator_sinkhorn(rho=rho_random, dim=[2, 2], max_iterations=20)
+    except RuntimeError as e:
+        expected_msg = "operator_sinkhorn did not converge within 20 iterations."
+        assert str(e) == expected_msg
+
+
+def test_operator_sinkhorn_near_zero_trace():
+    """Test if operator sinkhorn raises warnings on near zero output trace."""
+    rho_random = np.array(
+        [
+            [np.finfo(float).eps, 0, 0, 0],
+            [0, np.finfo(float).eps, 0, 0],
+            [0, 0, np.finfo(float).eps, 0],
+            [0, 0, 0, np.finfo(float).eps],
+        ]
+    )
+    expected_msg = "Final trace is near zero, but initial trace was not. Result may be unreliable."
+
+    with warnings.catch_warnings():
+        warnings.filterwarnings("error")
+        try:
+            operator_sinkhorn(rho=rho_random, dim=[2, 2])
+        except RuntimeWarning as warn_string:
+            assert str(warn_string) == expected_msg
+
+
+# Test outputs on maximally mixed and entangled states
+
+"""Test operator Sinkhorn on a maximally mixed bipartite state. Result should be invariant."""
+rho_mixed = np.eye(9)  # Identity matrix is a max-mixed state
+# The dimension `3` divides evenly into `9`, so no error should occur
+dim_mixed = [3]
+
+"""Test operator Sinkhorn on a maximally entangled bipartite state. Should be invariant."""
+# function should return the inintial state since it already satisfies the trace property
+u0 = bell(0)
+rho_ent = u0 @ u0.conj().T
+dim_ent = [2]
+
+params2 = [([rho_mixed, dim_mixed]), ([rho_ent, dim_ent])]
```

Comment:
Shouldn't these be parameters of `pytest.parameterize` instead of free-floating variables that live above a test? IOW, you can define them within the `parameterize` call. You can refer to other tests that use `pytest.parameterize` in the toqtio project for reference.